### PR TITLE
cmd/snap: iterate interface tab completion

### PR DESCRIPTION
**Pull request by chipaca**

### CODE REVIEW

File: cmd/snap/complete.go [hunk 1]
```diff
@@ -120,46 +212,73 @@ func (cps connectPlugSpec) Complete(match string) []flags.Completion {
 
 	var ret []flags.Completion
 
-	var snapPrefix, plugPrefix string
+	var prefix string
 	if len(parts) == 2 {
 		// The user typed the colon, means they know the snap they want;
 		// go with that.
```
Comment (stolowski) [inline cmd/snap/complete.go]: We also support colon followed by slot (i.e. snap name is empty), in which case we default to core, e.g.
`snap connect foosnap:home :home`
`snap disconnect :home`

It's a bit hard to see if that's supported by the code below? See ResolveConnect / ResolveDisconnect in the repo.

File: cmd/snap/complete.go [hunk 2]
```diff
@@ -120,46 +212,76 @@ func (cps connectPlugSpec) Complete(match string) []flags.Completion {
...
+					if plug.Snap == actualName && strings.HasPrefix(plug.Name, prefix) && spec.connFilter(len(plug.Connections)) {
+						// TODO: in the future annotate plugs that can take
+						// multiple connection sensibly and don't skip those even
+						// if they have connections already.
+						ret = append(ret, flags.Completion{Item: fmt.Sprintf("%s:%s", snapName, plug.Name), Description: "plug"})
+					}
+				}
+			}
+			if spec.slots {
+				if actualName == "" {
```
Comment (stolowski) [inline cmd/snap/complete.go]: Nitpick: if snapName == "" to make it consistent with the statement in line 246 above?

Comment (chipaca) [inline cmd/snap/complete.go]: counter-nitpick: that'll result in setting it to core when it's already core, in some situations :-)

Comment (stolowski) [inline cmd/snap/complete.go]: ah, good then!

File: cmd/snap/complete.go [hunk 3]
```diff
@@ -100,11 +100,103 @@ func (s keyName) Complete(match string) []flags.Completion {
 	return res
 }
 
+type disconnectSlotOrPlugSpec struct {
+	SnapAndName
+}
+
+func (dps disconnectSlotOrPlugSpec) Complete(match string) []flags.Completion {
+	spec := &interfaceSpec{
```
Comment (zyga) [inline cmd/snap/complete.go]: :heart: really nice design!

File: cmd/snap/complete.go [hunk 4]
```diff
@@ -100,11 +100,103 @@ func (s keyName) Complete(match string) []flags.Completion {
...
 type connectPlugSpec struct {
 	SnapAndName
 }
 
 func (cps connectPlugSpec) Complete(match string) []flags.Completion {
+	spec := &interfaceSpec{
+		SnapAndName:  cps.SnapAndName,
+		slots:        false,
+		plugs:        true,
+		connected:    false,
```
Comment (zyga) [inline cmd/snap/complete.go]: We may want to revise this when we model one-to-many and many-to-many connections more accurately.

File: cmd/snap/complete.go [hunk 5]
```diff
@@ -100,11 +100,103 @@ func (s keyName) Complete(match string) []flags.Completion {
...
+	SnapAndName
+}
+
+// TODO: look at what the previous arg is, and filter accordingly
+func (css connectSlotSpec) Complete(match string) []flags.Completion {
+	spec := &interfaceSpec{
+		SnapAndName:  css.SnapAndName,
+		slots:        true,
+		plugs:        false,
+		connected:    false,
```
Comment (zyga) [inline cmd/snap/complete.go]: We may want to revise this when we model one-to-many and many-to-many connections more accurately. A connected slot can still be connected to other plugs.

File: cmd/snap/complete.go [hunk 6]
```diff
@@ -120,46 +212,76 @@ func (cps connectPlugSpec) Complete(match string) []flags.Completion {
...
+			actualName := snapName
+			if spec.plugs {
+				if spec.connected && snapName == "" {
+					actualName = "core"
+				}
+				for _, plug := range ifaces.Plugs {
+					if plug.Snap == actualName && strings.HasPrefix(plug.Name, prefix) && spec.connFilter(len(plug.Connections)) {
+						// TODO: in the future annotate plugs that can take
+						// multiple connection sensibly and don't skip those even
+						// if they have connections already.
```
Comment (zyga) [inline cmd/snap/complete.go]: Thanks for adding this comment.

**Pull request merged**